### PR TITLE
Fix pipeline status when restarting the server #130

### PR DESCRIPTION
**Pull request by AvishrantsSh**
Fixed Issue #130.
Signed-off-by: Avishrant Sharma <avishrants@gmail.com>

### CODE REVIEW

File: scancodeio/settings/base.py [hunk 1]
```diff
@@ -222,6 +222,7 @@
 
 # Celery
 
+CELERY_BROKER_TRANSPORT_OPTIONS = {"socket_timeout": 5, "max_retries": 1}
```
Comment (pombredanne) [inline scancodeio/settings/base.py:225]: Could i make sense to use an environment variable here too?

File: scancodeio/wsgi.py [hunk 2]
```diff
@@ -33,6 +33,27 @@
...
+
 os.environ.setdefault("DJANGO_SETTINGS_MODULE", "scancodeio.settings")
 
+
+def get_celery_worker_status():
+    try:
+        i = app.control.inspect()
+        active = i.active()
+        scheduled = i.scheduled()
+        if active[list(active.keys())[0]] or scheduled[list(scheduled.keys())[0]]:
```
Comment (pombredanne) [inline scancodeio/wsgi.py:49]: Could there be a way to make this easier to read and understand? may be some intermediate variable? and is there a risk to use the first element? can this fail if the list are empty? Can they be empty? Is  there a difference between `list(active.keys())` and `list(active)` assuming `active` is some kind of mapping?
What are you trying to achieve here? get the first value of either one of two mappings?

File: scancodeio/wsgi.py [hunk 3]
```diff
@@ -33,6 +33,27 @@
...
+            return True
+    except OperationalError:
+        return False
+
+
+result = get_celery_worker_status()
+if not result:
+    Run.objects.filter(task_exitcode__isnull=True).update(task_exitcode=1)
```
Comment (pombredanne) [inline scancodeio/wsgi.py:57]: Can you elaborate what this does?

**Review by pombredanne — CHANGES_REQUESTED** [review]
Thanks! See my comments inline. Is there a way to get a test for this too? or is the setup too complex?

File: scancodeio/wsgi.py (hunk 2 above)
Comment (AvishrantsSh) [reply scancodeio/wsgi.py:49]: I will change variable names and try to make it more readable. When testing on my PC, `i.active()` returned a dictionary with one key of the format `celery@<-pc name->`, so I had to use the first key by using `list(active.keys())[0]`.  If no worker is avaliable, it simply returns a dictionary object like `{'celery@avishrant-Latitude-3500': []}`.

`can this fail if the list are empty?` - @pombredanne both active() and scheduled() will always work provided that the broker is reachable.

`What are you trying to achieve here?` - This code snippet simply checks if any job is running or scheduled. If it finds such a job, the Database won't be updated. Alternatively, if no job is running and any 'Run' database entry has `running`  task status, it simply updates that field, thereby solving the problem.

File: scancodeio/wsgi.py (hunk 3 above)
Comment (AvishrantsSh) [reply scancodeio/wsgi.py:57]: `Run.objects.filter(task_exitcode__isnull=True).update(task_exitcode=1)` - This command is executed when no worker is available at the moment when server is started. The filter extracts all entries having `task_exitcode = NULL`, which represents `Running` state,  and updates them to represent failure in execution `task_exitcode = 1`.

File: scancodeio/settings/base.py (hunk 1 above)
Comment (AvishrantsSh) [reply scancodeio/settings/base.py:225]: > Is there a way to get a test for this too? or is the setup too complex?

I'm not sure how to make a test case for this. Any ideas?

Comment (AvishrantsSh) [reply scancodeio/settings/base.py:225]: > Could i make sense to use an environment variable here too?

Yes, environment variables can be used to add Broker Configuration.

**Comment by pombredanne** [comment]
@tdruez @AvishrantsSh gentle ping... what's left to do to finish this?

**Review by pombredanne — APPROVED** [review]

**Review by pombredanne — CHANGES_REQUESTED** [review]
@tdruez your review is needed

**Comment by tdruez** [comment]
@pombredanne This is an incomplete solution, see https://github.com/nexB/scancode.io/issues/130#issuecomment-860731952 for details.
I think we should close this PR.

**Comment by tdruez** [comment]
Closing for now, see https://github.com/nexB/scancode.io/issues/130 for discussion.

**Pull request closed without merging**